File: src/types/service-program.ts
import type { EntityId, ServiceItemType, TimestampedEntity } from "./common";
import type { ProjectionContent } from "./projection";

export interface ServiceProgram extends TimestampedEntity {
  title: string;
  serviceDate?: string | null;
  notes?: string | null;
  isActive: boolean;
  items: ServiceItem[];
}

export interface ServiceProgramInput {
  title: string;
  serviceDate?: string;
  notes?: string;
  isActive?: boolean;
}

export interface ServiceItem extends TimestampedEntity {
  serviceProgramId: EntityId;
  itemType: ServiceItemType;
  title: string;
  linkedEntityId?: EntityId | null;
  customContentJson?: ProjectionContent | null;
  position: number;
}

export interface ServiceItemInput {
  serviceProgramId: EntityId;
  itemType: ServiceItemType;
  title: string;
  linkedEntityId?: EntityId;
  customContentJson?: ProjectionContent;
  position?: number;
}

File: src-tauri/src/models/service_program.rs
use serde::{Deserialize, Serialize};

use super::projection::ProjectionContent;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceProgram {
    pub id: i64,
    pub title: String,
    pub service_date: Option<String>,
    pub notes: Option<String>,
    pub is_active: bool,
    pub created_at: String,
    pub updated_at: String,
    pub deleted_at: Option<String>,
    pub items: Vec<ServiceItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceProgramInput {
    pub title: String,
    pub service_date: Option<String>,
    pub notes: Option<String>,
    pub is_active: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceItem {
    pub id: i64,
    pub service_program_id: i64,
    pub item_type: String,
    pub title: String,
    pub linked_entity_id: Option<i64>,
    pub custom_content_json: Option<ProjectionContent>,
    pub position: i64,
    pub created_at: String,
    pub updated_at: String,
    pub deleted_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceItemInput {
    pub service_program_id: i64,
    pub item_type: String,
    pub title: String,
    pub linked_entity_id: Option<i64>,
    pub custom_content_json: Option<ProjectionContent>,
    pub position: Option<i64>,
}
